luaui/Widgets/gui_top_bar.lua

```diff
@@ -1087,29 +1849,351 @@ function widget:GameStart()
+	local unp = unpack or table.unpack
+	--LogFrame("gameFrame: " .. gameFrame .. ", gameFrameFreq: " .. gameFrameFreq .. ", nowChecking: " .. nowChecking .. ", trackPosBase: " .. trackPosBase .. ", trackedNum: " .. trackedNum)
+
+	-- If we're supposed to draw the buildpower bar, do some calculations.
+	-- Skip frames between calculations _unless_ we have too many builders to do them all in one frame, in which case nowChecking will be positive when the previous frame didn't finish.
+	if config.drawBPBar and ((gameFrame % gameFrameFreq == 0) or nowChecking > 0) then
+		gameStarted = true -- TODO: is this needed?
+		--LogFrame("Processing BPBar calculations")
+
+		-- Log initial cache values
+		local cacheTotalReservedBP = 0
+		local cacheTotallyUsedBP = 0
+
+		local usedBPMetalExpense = 0
+		local usedBPEnergyExpense = 0
+		local buildingBP = 0 -- how much BP is actively building, regardless of how stalled it is?
+		local nonStalledBuildingBP = 0 -- how much BP is actively building and not stalled?
+
+		local unitDefsBeingBuilt = {}
+		local builderStates = {}
+		local unitsReservedBP = {}
+
+		--LogFrame("Starting trackedBuilders loop")
+
+		-- Initialisieren des unitIterator, falls dies der erste Frame ist oder die Coroutine beendet ist
+		if not unitIterator or coroutine.status(unitIterator) == "dead" then
+			unitIterator = coroutine.create(function()
+				for unitID, unitData in pairs(trackedBuilders) do
+					coroutine.yield(unitID, unitData)
+				end
+			end)
+		end
+
+		-- Verwenden des Iterators, um Einheiten für diesen Frame zu verarbeiten
+		for i = 1, unitsPerFrame do
+			local success, unitID, unitData = coroutine.resume(unitIterator)
+			if not success or not unitID then
+				break -- Keine weiteren Einheiten mehr zum Verarbeiten oder Fehler
+			end
+
+			--LogFrame("Processing unitID: " .. tostring(unitID) .. ", nowChecking: " .. nowChecking .. ", trackPosBase: " .. trackPosBase)
+			local currentUnitBP, unitIsBuilt, unitDefID, unitTeamID = unp(unitData)
+			--LogFrame("Unit is built, performing additional checks")
+			local unitExists, foundActivity, builtUnitDefID, mayBeBuilding, guardedUnitID = findBPCommand(unitID, unitDefID, {CMD.REPAIR, CMD.RECLAIM, CMD.CAPTURE, CMD.GUARD})
+			--LogFrame("Unit exists: " .. tostring(unitExists) .. ", foundActivity: " .. tostring(foundActivity) .. ", builtUnitDefID: " .. tostring(builtUnitDefID) .. ", mayBeBuilding: " .. tostring(mayBeBuilding) .. ", guardedUnitID: " .. tostring(guardedUnitID))
+
+			if not unitExists then
+				--LogFrame("Unit no longer exists, untracking unitID: " .. unitID)
+				
+				UntrackUnit(unitID)
+			elseif not unitIsBuilt then
+				break
+
+			elseif not foundActivity then
+				--LogFrame("Unit not found doing any activity, marking as idle")
+				
+				builderStates[unitID] = { false, builtUnitDefID, guardedUnitID, currentUnitBP }
```

Comment:
If this is already created, can we edit the values rather than creating a new table, to avoid garbage generation?

Also L1954

Reply:
Mr. "Major Clean up"'s work, again. 

This will take time to get rid of, again.

Reply:
At least for me and my skill level

Reply:
done!